pennylane/transforms/mitigate.py

```diff
@@ -12,15 +12,346 @@
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†─╰X──RZ(5.0)─┤ ╰<Z@Z@Z>
+
+    Setting ``scale_factor = 3`` results in the folded circuit :math:`U (U^\dagger U)`.
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 3)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──RX(3.0)†───────────────╭●─────────RX(0.0)†──RX(0.0)─╭●──RX(3.0)──────────┤╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)───RY(4.0)†─╭●──╰X†────────RY(1.0)†──RY(1.0)─╰X─╭●────────RY(4.0)─┤├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†──RZ(2.0)†──RZ(2.0)──────────────╰X────────RZ(5.0)─┤╰<Z@Z@Z>
+    """
+
+    if scale_factor < 1.0:
+        raise AttributeError("scale_factor must be >= 1")
+    assert scale_factor >= 1.0
+
+    # TODO: simplify this - currently just a workaround, problem is when tape contains adjoint(op)
+    def qfunc(op):
+        copy(op).queue()
+
+    # Generate base_circuit without measurements
+    # Treat all circuits as lists of operations, build new tape in the end
+
+    base_ops = circuit.expand().copy(copy_operations=True).operations
+
+    def _divmod(a, b):
+        """Performs divmod but in an all-interface compatible manner"""
+        out1 = qml.math.floor(a / b)
+        out2 = a - out1 * b
+        return int(out1), int(out2)
+
+    num_global_folds, fraction_scale = _divmod(scale_factor - 1, 2)
+
+    n_ops = len(base_ops)
+    num_to_fold = int(round(fraction_scale * n_ops / 2))
+
+    # Create new_circuit from folded list
+    with QuantumTape() as new_circuit:
+        # Original U
+        for op in base_ops:
+            qfunc(op)
+
+        # Folding U => U (U^H U)**n.
+        for _ in range(int(num_global_folds)):
+            for op in base_ops[::-1]:
+                adjoint(qfunc)(op)
+
+            for op in base_ops:
+                qfunc(op)
+
+        # Remainder folding U => U (U^H U)**n (L_d^H .. L_s^H) (L_s .. L_d)
+        for i in range(n_ops - 1, n_ops - num_to_fold - 1, -1):
+            adjoint(qfunc)(base_ops[i])
+
+        for i in range(n_ops - num_to_fold, n_ops):
+            qfunc(base_ops[i])
+
+        # Append measurements
+        for meas in circuit.measurements:
+            apply(meas)
+
+    # return [new_circuit], lambda x: x
+    return new_circuit
+
+
+# TODO: make this a pennylane.math function
+def _polyfit(x, y, order):
+    """Brute force implementation of polynomial fit"""
+    lhs = qml.math.vander(x, order + 1)
+    rhs = qml.math.stack(y)  # [qml.math.stack(i) for i in y]
+
+    # scale lhs to improve condition number and solve
+    scale = qml.math.sum(qml.math.sqrt((lhs * lhs)), axis=0)
+    lhs /= scale
+
+    # Compute coeffs:
+    # This part is typically done using a lstq solver, do it with the penrose inverse by hand:
+    # i.e. coeffs = (X.T @ X)**-1 X.T @ y see https://en.wikipedia.org/wiki/Polynomial_regression
+    c = qml.math.linalg.pinv(qml.math.transpose(lhs) @ lhs)
```

Comment:
Do you have the issue at hand? I guess you are not referring to https://github.com/PennyLaneAI/pennylane/issues/299 ? I can make some checks, but would need more context what to look for.

Reply:
I think it might be worth looking into anyway because it is probable faster since `pinv` is doing an SVD. My impression from some testing is that `linalg.solve` is not differentiable?

```python
x = np.linspace(0,10,10, requires_grad=False)
y = x**2 + x + 1.
y.requires_grad=True

def func(x, y):
    X = qml.math.vander(x, 2 + 1)
    yy = qml.math.stack(y)  # [qml.math.stack(i) for i in y]

    # scale X to improve condition number and solve
    scale = qml.math.sum(qml.math.sqrt((X * X)), axis=0)
    X /= scale

    # another alternative: rewrite as (X.T @ X) coeffs = X.T @ y and use ``linalg.solve``
    lhs = qml.math.cast_like(qml.math.transpose(X) @ X, yy)
    rhs = qml.math.dot(qml.math.transpose(X), yy)
    c = qml.math.linalg.solve(lhs, rhs)

    c = qml.math.transpose(qml.math.transpose(c) / scale)
    return c

>>> qml.grad(func)(x, y)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/qottmann/Qottmann/Xanadu/Mitigate_fold_global_ZNE.ipynb Cell 11' in <cell line: 1>()
----> 1 qml.grad(func)(x, y)

File ~/Qottmann/Xanadu/pennylane/pennylane/_grad.py:115, in grad.__call__(self, *args, **kwargs)
    112     self._forward = self._fun(*args, **kwargs)
    113     return ()
--> 115 grad_value, ans = grad_fn(*args, **kwargs)
    116 self._forward = ans
    118 return grad_value

File ~/anaconda3/envs/pennylane/lib/python3.8/site-packages/autograd/wrap_util.py:20, in unary_to_nary.<locals>.nary_operator.<locals>.nary_f(*args, **kwargs)
     18 else:
     19     x = tuple(args[i] for i in argnum)
---> 20 return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)

File ~/Qottmann/Xanadu/pennylane/pennylane/_grad.py:133, in grad._grad_with_forward(fun, x)
    127 @staticmethod
    128 @unary_to_nary
    129 def _grad_with_forward(fun, x):
    130     """This function is a replica of ``autograd.grad``, with the only
    131     difference being that it returns both the gradient *and* the forward pass
    132     value."""
--> 133     vjp, ans = _make_vjp(fun, x)
...
    392 extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 393 r = gufunc(a, b, signature=signature, extobj=extobj)
    395 return wrap(r.astype(result_t, copy=False))

TypeError: No loop matching the specified signature and casting was found for ufunc solve1
```

Reply:
Ah! I wasn't aware that `solve` wasn't differentiable, interesting :( In the case of the metric tensor, it is used inside the QNG optimizer, so that would be why we never noticed the lack of differentiability.